File: DotNet/PopulationFitness/PopulationFitness/Models/FastMaths/FastMaths.cs
namespace PopulationFitness.Models.FastMaths
{
    public class FastMaths
    {
        /**
         * Optimised version of power
         * @param baseValue
         * @param power
         * @return base raised to the power
         */
        public static double Pow(double baseValue, long power)
        {
            double result = 1;
            while (power > 0)
            {
                if ((power & 1) == 1)
                {
                    result *= baseValue;
                }
                power >>= 1;
                baseValue *= baseValue;
            }
            return result;
        }
    }
}

File: Scala/PopulationFitness/src/main/scala/uk/edu/populationfitness/models/maths/FastMaths.scala
package uk.edu.populationfitness.models.maths

object FastMaths {
  def pow(base: Double, power: Long): Double = {
    var result = 1.0
    var currentPower = power
    var currentBase = base
    while ( currentPower > 0) {
      if ((currentPower & 1) == 1) result *= currentBase
      currentPower >>= 1
      currentBase *= currentBase
    }
    result
  }
}
